city/repositories/city_repository.py

```diff
@@ -0,0 +1,52 @@
+from sqlalchemy import select
+from sqlalchemy.ext.asyncio import AsyncSession
+
+from city import schemas
+from db import models
+
+
+class CityRepository:
+    def __init__(self, session: AsyncSession):
+        self.session = session
+
+    async def get_cities_list(self):
+        query = select(models.City)
+        cities_list = await self.session.execute(query)
+        return [city[0] for city in cities_list.fetchall()]
+
+    async def get_city_by_id(self, city_id: int):
+        query = select(models.City).where(models.City.id == city_id)
+        result = await self.session.execute(query)
+        return result.scalar_one_or_none()
+
+    async def update_city_by_id(
+        self,
+        city_id: int,
+        city: schemas.CityUpdate,
+    ) -> models.City | None:
+        db_city = await self.get_city_by_id(city_id)
+        if not db_city:
+            return None
+        db_city.name = city.name
+        db_city.additional_info = city.additional_info
+        await self.session.commit()
+        await self.session.refresh(db_city)
+        return db_city
+
+    async def create_city(self, city: schemas.CityCreate):
+        db_city = models.City(
+            name=city.name,
+            additional_info=city.additional_info
+        )
+        self.session.add(db_city)
+        await self.session.commit()
+        await self.session.refresh(db_city)
+        return db_city
+
+    async def delete_city_by_id(self, city_id: int) -> bool:
+        db_city = await self.get_city_by_id(city_id)
+        if not db_city:
+            return False
+        await self.session.delete(db_city)
+        await self.session.commit()
+        return True
```

Comment:
Tight Coupling with ORM Models
The Service class is tightly coupled with the ORM models (models.DBCity, models.DBTemperature). This makes it difficult to test the service in isolation because it directly depends on the database models.
Improvement: Use Data Transfer Objects (DTOs) or schemas for input and output. This way, the service methods can return plain Python objects or schemas, which are easier to mock and test.

 Lack of Transaction Management
The create_city, update_city, and delete_city methods commit the session after each operation. If you need to perform multiple operations within a single transaction, this approach could lead to issues.
Improvement: Implement a transaction management strategy, such as using a context manager that manages the session's lifecycle and commits or rolls back based on success or failure.

Error Handling
There's no error handling in any of the methods. If something goes wrong (e.g., the city doesn't exist, a database constraint is violated, etc.), it could cause unhandled exceptions.
Improvement: Implement error handling with appropriate custom exceptions. This will make the service more robust and predictable.

Session Management in create_temperatures
The create_temperatures method makes multiple commit() calls within a loop, which could lead to partial updates if something fails in the middle.

Reply:
Error handling in service layer (city_service, temperatures_service)